### PR TITLE
Adding orog gsl to orog mask tools

### PR DESCRIPTION
**Pull request by mdtoy**
Added directory "orog_gsl.fd" to "orog_mask_tools.fd".  Contains Fortran code which generates "oro_data" netCDF files needed for GSL subgrid-scale orographic drag suite.

### CODE REVIEW

**Comment by GeorgeGayno-NOAA** [comment]
Two initial comments:

- We now require a unit/regression test for all new codes added to the repository.  We can help with creating a test.
- Under ./driver_scripts, users can set up their own grids by running a series of steps that create the geo-referencing fields, the land mask and terrain/GWD; invoke the topo smoothing, etc.  Can your code be made an option for users to choose how to set the GWD fields?

**Comment by mdtoy** [comment]
- Sounds good regarding the unit/regression test.  Let me know what the first step towards this would be.
- I believe my code can be made an option for creating the GWD fields.  The user would specify whether they want to have GSL_drag fields or not.  The call to "orog_gsl" would be very similar to that for the existing "orog" call in ./ush/fv3gfs_make_orog.sh.

File: sorc/orog_mask_tools.fd/orog_gsl.fd/CMakeLists.txt
Comment (GeorgeGayno-NOAA) [inline sorc/orog_mask_tools.fd/orog_gsl.fd/CMakeLists.txt:20]: Your program is self-contained.  So there is no need to map in the bacio, w3nco, ip or sp libraries.  And if you are not using threads, you can remove the reference to OpenMP_Fortran.

**Comment by GeorgeGayno-NOAA** [comment]
We are also converting our repository to use doxygen.  Please add at least minimal doxygen to the top of your modules.  See ./sorc/chgres_cube.fd for examples.

**Comment by edwardhartnett** [comment]
Yes please on automated testing and doxygen documentation! Ask for help if you need it!

**Comment by GeorgeGayno-NOAA** [comment]
@mdtoy I see you are using the F77 netcdf library calls.  Can you convert to use the F90 version?  @edwardhartnett you are our netcdf expert.  Do you agree?

**Comment by edwardhartnett** [comment]
The F77 netCDF calls will continue to work, and do the same things as the F90 calls.

That said, I do think you should use the netCDF F90 API for new codes.

File: sorc/orog_mask_tools.fd/orog_gsl.fd/CMakeLists.txt
Comment (mdtoy) [reply sorc/orog_mask_tools.fd/orog_gsl.fd/CMakeLists.txt:20]: I modified CMakeLists.txt according to your suggestions.

**Comment by mdtoy** [comment]
I added doxygen comments to the oro_gsl.fd/*.f90 files.

**Comment by mdtoy** [comment]
I'm happy to convert the netCDF calls to F90 API.  I'm not a programming expert, and I actually didn't know I'd been using F77 calls in the first place.  Would the conversion be line-by-line?  Do you know of a good reference for dummies that would show the conversion from F77 to F90?

**Comment by edwardhartnett** [comment]
I don't know of any better resource than the netcdf-fortran docs:
https://www.unidata.ucar.edu/software/netcdf/docs-fortran/

There is usually a 1 to 1 correspondence of functions between F90 and F77, but in the case of F90 there are some functions that use overloading so that one function in F90 maps to multiple functions in F77.

**Comment by GeorgeGayno-NOAA** [comment]
> I'm happy to convert the netCDF calls to F90 API. I'm not a programming expert, and I actually didn't know I'd been using F77 calls in the first place. Would the conversion be line-by-line? Do you know of a good reference for dummies that would show the conversion from F77 to F90?

Under ./sorc/chgres_cube.fd  You can look at write_data.F90 (for writing files); and thompson_mp_climo_data.F90 or iinput_data.F90 (for reading files).

**Comment by mdtoy** [comment]
I just updated the netCDF calls to the F90 API.

File: sorc/orog_mask_tools.fd/orog_gsl.fd/gsl_oro_data.f90
Comment (edwardhartnett) [inline sorc/orog_mask_tools.fd/orog_gsl.fd/gsl_oro_data.f90:21]: Please use the doxygen author and date tags.

File: sorc/orog_mask_tools.fd/orog_gsl.fd/gsl_oro_data.f90
Comment (edwardhartnett) [inline sorc/orog_mask_tools.fd/orog_gsl.fd/gsl_oro_data.f90:12]: Can you make these proper lists? Just use a hyphen, like this:
```
- var
- con
- etc.
```

File: sorc/orog_mask_tools.fd/orog_gsl.fd/module_gsl_oro_data_lg_scale.f90
Comment (edwardhartnett) [inline sorc/orog_mask_tools.fd/orog_gsl.fd/module_gsl_oro_data_lg_scale.f90:4]: Full name instead of user id for author please.

**Review by edwardhartnett — CHANGES_REQUESTED** [review]
Do we have tests for this code?

**Comment by mdtoy** [comment]
Are you asking me if there are tests?  I did my own tests, but is there a formal testing platform?  If so, I'm not familiar with this.  Please let me know what I would need to do to get on board with testing.

**Comment by mdtoy** [comment]
George,

I tested the 'orog_gsl' executable within the regional_workflow and an issue that came up was that if the 'Cxxx_grid.tilexx.halo4.nc' file is not present, 'orog_gsl' runs anyway and creates netCDF orog_data files containing zeros.  I need to add a test for the presence of the 'grid' data.  In the program 'mtnlm7_oclsm.f' (part of orog.fd), there is a test that calls "errexit(4)", as in:
         inquire(file=trim(OUTGRID), exist=fexist)
         if(.not. fexist) then
            print*, "file "//trim(OUTGRID)//" does not exist"
            CALL ERREXIT(4)
         endif
Should I put similar lines of code in my Fortran code?  Is this meet Fortran 90 standards?  Also, does it matter what the exit code is (in the case above, it is "4")?

Thanks,
Mike

**Comment by GeorgeGayno-NOAA** [comment]
> George,
> 
> I tested the 'orog_gsl' executable within the regional_workflow and an issue that came up was that if the 'Cxxx_grid.tilexx.halo4.nc' file is not present, 'orog_gsl' runs anyway and creates netCDF orog_data files containing zeros. I need to add a test for the presence of the 'grid' data. In the program 'mtnlm7_oclsm.f' (part of orog.fd), there is a test that calls "errexit(4)", as in:
> inquire(file=trim(OUTGRID), exist=fexist)
> if(.not. fexist) then
> print*, "file "//trim(OUTGRID)//" does not exist"
> CALL ERREXIT(4)
> endif
> Should I put similar lines of code in my Fortran code? Is this meet Fortran 90 standards? Also, does it matter what the exit code is (in the case above, it is "4")?
> 
> Thanks,
> Mike

Yes, please add error handling for the reads and writes of any files.  We don't require a specific exit code number as long as it is non-zero.  But we do want any error that halts execution to have a message that indicates a "fatal error".  For example, see:
https://github.com/NOAA-EMC/UFS_UTILS/blob/develop/sorc/chgres_cube.fd/utils.f90#L21

When reading/writing netcdf, I check all calls to the library for errors.  For example, see: https://github.com/NOAA-EMC/UFS_UTILS/blob/develop/sorc/chgres_cube.fd/input_data.F90#L1525

Frequent error checks are useful when developing code, and helps users know what went wrong.

**Comment by mdtoy** [comment]
I modified ush/fv3gfs_driver_grid.sh and added a new script ush/fv3gfs_make_orog_gsl.sh, which tests the new sorc/orog_mask_tools.fd/orog_gsl.fd orographic statistics maker for the GSL gravity wave drag physics parameterization.

My own tests successfully generated the "C*_oro_data_ss*nc" and "C*_oro_data_ls*nc" files on Hera at:
/scratch1/BMC/gsd-fv3-dev/NCEPDEV/stmp3/Michael.Toy/my_grids.

I'm not sure why the automatic Jasper tests are failing.

Please advise on the next steps to getting this PR approved.  Thank you.

**Comment by GeorgeGayno-NOAA** [comment]
> I modified ush/fv3gfs_driver_grid.sh and added a new script ush/fv3gfs_make_orog_gsl.sh, which tests the new sorc/orog_mask_tools.fd/orog_gsl.fd orographic statistics maker for the GSL gravity wave drag physics parameterization.
> 
> My own tests successfully generated the "C*_oro_data_ss_nc" and "C__oro_data_ls*nc" files on Hera at:
> /scratch1/BMC/gsd-fv3-dev/NCEPDEV/stmp3/Michael.Toy/my_grids.
> 
> I'm not sure why the automatic Jasper tests are failing.
> 
> Please advise on the next steps to getting this PR approved. Thank you.

I see the failing tests.  I don't think you are doing anything wrong.  Github has been acting strange lately.

**Comment by mdtoy** [comment]
Okay, that’s good to know.  Let me know what happens next.  Thanks.

Best regards,
Mike

**Comment by JeffBeck-NOAA** [comment]
@GeorgeGayno-NOAA, @edwardhartnett, do you see anything more @mdtoy needs to do related to this PR?  I'm not sure how to rectify the automatic tests.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA, @edwardhartnett, do you see anything more @mdtoy needs to do related to this PR? I'm not sure how to rectify the automatic tests.

Since this is a new option to the grid generation process, I would like to add another test to ./reg_tests/grid_gen.  I will help with that.  Let me try a first hack.

**Comment by GeorgeGayno-NOAA** [comment]
@mdtoy I was able to get your branch to compile on Hera.  I created regional_esg and global uniform grids.  A couple questions:  

- I see three sets of 'orog' files.  Is that expected?  

- Your code only creates GWD fields, correct?  So the land mask, mean terrain and filtered terrain will continue to come from the 'orog.fd' program?  And, if the user chooses the new fields, the GWD fields from the 'orog' program will not be used?

**Comment by mdtoy** [comment]
George,

Yes, there should be three “oro_data” files, e.g., Cxxx_oro_data.tilex.nc, Cxxx_oro_data.tilex_ls.nc and Cxxx_oro_data_ss.tilex.nc.  The first one is the standard orog data including landmask, terrain, and gravity wave drag fields from the “orog.fd” program.  The last two are the ones my code creates, and yes, they only contain GWD fields.  Yes, if the user chooses the GSL drag suite, the GWD fields from the ‘orog’ program will not be used, but the terrain and landmask info will.

— Mike

File: ush/fv3gfs_driver_grid.sh
Comment (GeorgeGayno-NOAA) [inline ush/fv3gfs_driver_grid.sh:105]: Your input terrain datasets should really be in 'fix_orog' with the others, not in fix_am.  I will get with @KateFriedman-NOAA to move them.  Then you can get rid of the topo_am variable.

**Comment by GeorgeGayno-NOAA** [comment]
Can we make the creation of these additional file an option?

File: sorc/orog_mask_tools.fd/orog_gsl.fd/README
Comment (GeorgeGayno-NOAA) [inline sorc/orog_mask_tools.fd/orog_gsl.fd/README:5]: Can you list the orientations here and in the prologs?  1 is N?  2 is SW?

**Comment by GeorgeGayno-NOAA** [comment]
I am seeing a lot of doxygen warnings as well.  Let me try to clear those for you.

**Comment by GeorgeGayno-NOAA** [comment]
@mdtoy I cleared the doxygen warnings and committed the changes to your branch.  Search through the source code for "???" and  replace with a definition.

**Comment by mdtoy** [comment]
Thanks for your help with the doxygen commenting.  I added descriptions in the appropriate places.

File: ush/fv3gfs_driver_grid.sh
Comment (mdtoy) [reply ush/fv3gfs_driver_grid.sh:105]: Sounds good.

**Comment by mdtoy** [comment]
@GeorgeGayno-NOAA  in regard to your comment "Can we make the creation of these additional file an option?".  If you're referring to the oro_data_ls and oro_data_ss files created by orog_gsl, we made their creation (i.e., execution of "orog_gsl") an option based on which Gravity Wave Drag physics package is used in CCPP.  So yes, it can be an option.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA in regard to your comment "Can we make the creation of these additional file an option?". If you're referring to the oro_data_ls and oro_data_ss files created by orog_gsl, we made their creation (i.e., execution of "orog_gsl") an option based on which Gravity Wave Drag physics package is used in CCPP. So yes, it can be an option.

Ok.  So can we add a flag in the ./driver_scripts so the user can create these extra "orog" files or not.  Users who create a grid and want to run the model with the current GWD might get confused by the extra files.

**Comment by GeorgeGayno-NOAA** [comment]
@mdtoy I will work with you to create a new regression test.  Which is the best grid to use?  Global uniform?  Regional ESG?  Does not matter?

**Comment by mdtoy** [comment]
@GeorgeGayno-NOAA Do you have a preference for the flag in ./driver_scripts?  What do you think of the following?  The new flag would be called "make_gsl_orog".


**./driver_scripts/driver_grid.hera.sh:**

#-----------------------------------------------------------------------
# Set grid specs here.
#-----------------------------------------------------------------------

export gtype=uniform           # 'uniform', 'stretch', 'nest', 
                               # 'regional_gfdl', 'regional_esg'

export make_gsl_orog = true  # or false


**./ush/fv3gfs_driver_grid.sh**
if [ $make_gsl_orog = true ] ; then
     calls to fv3gfs_make_orog_gsl.sh
fi

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA Do you have a preference for the flag in ./driver_scripts? What do you think of the following? The new flag would be called "make_gsl_orog".
> 
> **./driver_scripts/driver_grid.hera.sh:**
> 
> #-----------------------------------------------------------------------
> 
> # Set grid specs here.
> #-----------------------------------------------------------------------
> 
> export gtype=uniform # 'uniform', 'stretch', 'nest',
> # 'regional_gfdl', 'regional_esg'
> 
> export make_gsl_orog = true # or false
> 
> **./ush/fv3gfs_driver_grid.sh**
> if [ $make_gsl_orog = true ] ; then
> calls to fv3gfs_make_orog_gsl.sh
> fi

"make_gsl_orog" is fine.  I would make the default value 'false'.  If you add the flag to all of the ./driver_scripts, I can test on the machines you do not have access to.

**Comment by GeorgeGayno-NOAA** [comment]
For the regression test, I chose a regional ESG grid.  The baseline data - against which all test output will be compared - is here on Hera: /scratch1/NCEPDEV/da/George.Gayno/noscrub/reg_tests/grid_gen/baseline_data/regional_gsl_gwd

Please give it a quick check.  

I will also create a new regression test script.  Would you mind if pushed any changes to your branch?

**Comment by mdtoy** [comment]
Coincidentally, I ran the same case (regional ESG), and your baseline data looks good.

Yes, you may push changes to my branch.

**Comment by mdtoy** [comment]
@GeorgeGayno-NOAA I tested out the regression test on Hera, and the "regional.gsl.gwd" test didn't produce the C772_oro_data_ls.tile7.halo0.nc and C772_oro_data_ss.tile7.halo0.nc files.
Test directory:
/scratch1/BMC/gsd-fv3-dev/NCEPDEV/stmp3/Michael.Toy/reg_tests.grid.1st_try/regional.gsl.gwd/C772

Baseline data:
/scratch1/NCEPDEV/da/George.Gayno/noscrub/reg_tests/grid_gen/baseline_data/regional.gsl.gwd

I'm looking at the scripts to find out why this is.

**Comment by mdtoy** [comment]
@GeorgeGayno-NOAA It looks like you already found the issue.  Line 56 of ./ush/fv3gfs_driver_grid.sh was missing a "$" sign in front of {make_gsl_orog:-false}.  I'm putting it in and will retest.

**Comment by mdtoy** [comment]
@GeorgeGayno-NOAA My regression test is failing.  One of the fields of the C772_oro_data_ls.tile7.halo0.nc files are slightly different from your baseline.  Can you direct me to the exec/orog_gsl executable that you used to generate the baselines?  I'd like to see if I can reproduce your result with it.  Thanks.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA My regression test is failing. One of the fields of the C772_oro_data_ls.tile7.halo0.nc files are slightly different from your baseline. Can you direct me to the exec/orog_gsl executable that you used to generate the baselines? I'd like to see if I can reproduce your result with it. Thanks.

Yes, I am getting that failure as well.  I will point you to the script I used to generate the baseline files.

**Comment by GeorgeGayno-NOAA** [comment]
@mdtoy To generate the baseline files, I used ./driver_scripts/driver_grid.hera.sh.  I only made these changes:

```
 #-----------------------------------------------------------------------
 # Driver script to create a cubic-sphere based model grid on Hera.
@@ -69,9 +69,9 @@ module list
 # Set grid specs here.
 #-----------------------------------------------------------------------

-export gtype=uniform           # 'uniform', 'stretch', 'nest',
+export gtype=regional_esg           # 'uniform', 'stretch', 'nest',
                                # 'regional_gfdl', 'regional_esg'
-export make_gsl_orog=false     # 'true' if user needs 'oro' files for GSL
+export make_gsl_orog=true     # 'true' if user needs 'oro' files for GSL
                                # orographic drag suite
```

The regression test runs the same scripts in ./ush.  So I am puzzled as to what is happening.  I don't see any errors in the log files.  It is only the "orog_ls" file that is different.  Both the regression test and driver script use 24 threads.  Is your code threaded?

**Comment by mdtoy** [comment]
@GeorgeGayno-NOAA No, my code is not threaded.

**Comment by GeorgeGayno-NOAA** [comment]
> > @GeorgeGayno-NOAA My regression test is failing. One of the fields of the C772_oro_data_ls.tile7.halo0.nc files are slightly different from your baseline. Can you direct me to the exec/orog_gsl executable that you used to generate the baselines? I'd like to see if I can reproduce your result with it. Thanks.
> 
> Yes, I am getting that failure as well. I will point you to the script I used to generate the baseline files.

I am getting the same difference on WCOSS-Dell.  It is the 'ol1' field.

**Comment by mdtoy** [comment]
@GeorgeGayno-NOAA This is baffling.  I checked the code and there isn't anything special about the "ol1" field.  Are you using the same executable "exec/orog_gsl" (not the driver script) for generating the baseline files as well as the test files?

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA This is baffling. I checked the code and there isn't anything special about the "ol1" field. Are you using the same executable "exec/orog_gsl" (not the driver script) for generating the baseline files as well as the test files?

I turned on bound checking for Intel (-check bounds).  It crashed here:

```
forrtl: severe (408): fort: (3): Subscript #2 of the array ZS has value 0 which is less than the lower bound of 1

Image              PC                Routine            Line        Source
orog_gsl           000000000043C800  Unknown               Unknown  Unknown
orog_gsl           00000000004189E4  gsl_oro_data_lg_s         689  module_gsl_oro_data_lg_scale.f90
orog_gsl           000000000042E630  MAIN__                     77  gsl_oro_data.f90
orog_gsl           0000000000409D1E  Unknown               Unknown  Unknown
libc-2.17.so       00002B2329532545  __libc_start_main     Unknown  Unknown
orog_gsl           0000000000409C29  Unknown               Unknown  Unknown
```

Is that a clue?

**Comment by mdtoy** [comment]
> > @GeorgeGayno-NOAA This is baffling. I checked the code and there isn't anything special about the "ol1" field. Are you using the same executable "exec/orog_gsl" (not the driver script) for generating the baseline files as well as the test files?
> 
> I turned on bound checking for Intel (-check bounds). It crashed here:
> 
> ```
> forrtl: severe (408): fort: (3): Subscript #2 of the array ZS has value 0 which is less than the lower bound of 1
> 
> Image              PC                Routine            Line        Source
> orog_gsl           000000000043C800  Unknown               Unknown  Unknown
> orog_gsl           00000000004189E4  gsl_oro_data_lg_s         689  module_gsl_oro_data_lg_scale.f90
> orog_gsl           000000000042E630  MAIN__                     77  gsl_oro_data.f90
> orog_gsl           0000000000409D1E  Unknown               Unknown  Unknown
> libc-2.17.so       00002B2329532545  __libc_start_main     Unknown  Unknown
> orog_gsl           0000000000409C29  Unknown               Unknown  Unknown
> ```
> 
> Is that a clue?

Yes, that's it.  I fixed the code to avoid out-of-bounds arrays.  Bounds-check passes now.  Thank you for finding this.
I'm about to push my changes to my fork.

**Comment by GeorgeGayno-NOAA** [comment]
I updated the regression test driver scripts for Dell and Cray.  May I push them to your branch?

**Comment by mdtoy** [comment]
Yes.

**Comment by GeorgeGayno-NOAA** [comment]
I updated the baseline data on Hera.  The regression test is passing now.  I will update the baseline data on our other machines.

**Comment by GeorgeGayno-NOAA** [comment]
@mdtoy I have the regression test working on all our officially supported machines.  And the grid driver scripts are working.  So we are getting closer to merging.  

@edwardhartnett if additional automated tests are required, we may need some of your or Kyle's help.

**Comment by mdtoy** [comment]
That’s great news.  Thank you.

**Comment by JeffBeck-NOAA** [comment]
Hi @mdtoy, @GeorgeGayno-NOAA, just checking in to see whether this PR is ready?  Thanks!

**Comment by GeorgeGayno-NOAA** [comment]
> Hi @mdtoy, @GeorgeGayno-NOAA, just checking in to see whether this PR is ready? Thanks!

That depends on whether automated tests will be required. If not, then it is ready to merge. Unfortunately, I can't help with automated tests as I am still learning how they work.

**Comment by edwardhartnett** [comment]
We need unit testing for all new additions to UFS_UTILS.

In this PR you add module gsl_oro_data_lg_scale. In that module there is a subroutine calc_gsl_oro_data_lg_scale(tile_num,res_indx,halo).

We need a test program which calls this subroutine with known inputs and then checks the output for correctness. The test program should return 0 for success and non-zero for failure. See this PR for an MPI-based example: https://github.com/NOAA-EMC/UFS_UTILS/pull/405.

**Comment by mdtoy** [comment]
I thought we already completed this step (George Gayno and I) a couple of weeks ago with the addition regression tests with calls to the orog_gsl subroutines.

**Comment by edwardhartnett** [comment]
How did we verify the output of the regression tests? In other words, why do we think it is correct?

**Comment by mdtoy** [comment]
I created the algorithms and the code for generating the orographic data used by the GSL drag suite.  The files created by the RTs are identical to my baseline files.

**Comment by edwardhartnett** [comment]
Excellent. How do we know the baseline files are correct?

**Comment by mdtoy** [comment]
Can we get on a conference call with George so we can discuss this more efficiently?  I’m free after 3:30pm EDT today, or almost anytime tomorrow (Tuesday).

**Comment by edwardhartnett** [comment]
Absolutely. What is your email address and I will set up a meeting.

**Comment by mdtoy** [comment]
It’s ***@***.*** ***@***.***>

Thanks.

**Comment by edwardhartnett** [comment]
@mdtoy if that was supposed to be intelligable, I'm missing something. Why don't you set up the meeting? edward.hartnett@noaa.gov.

**Comment by JeffBeck-NOAA** [comment]
GitHub appears to be blocking email addresses for some reason.  Mike's email is michael.toy@noaa.gov.

**Comment by GeorgeGayno-NOAA** [comment]
@mdtoy Please merge the latest updates from 'develop' to your branch.  Then I will do some final builds and make sure the regression tests run OK on our five officially supported machines.

**Comment by mdtoy** [comment]
Done.

**Comment by edwardhartnett** [comment]
I do not believe that this code is generating correct science results. 

There are about 2500 lines of new Fortran code in this PR.

We (myself, @GeorgeGayno-NOAA , @mdtoy ) had a meeting to discuss the level of testing that this code has been subjected to:
1 - @mdtoy manually checked results at about eight different times during development. 
2 - The regression test output was generated and scanned for infinities.
3 - The code was run within the model without causing any problems.
4 - The output was then accepted and now forms the basis of our regression tests.

We have no tests that check or demonstrate that the actual numbers generated are correct. We have no tests for boundary or error conditions. 

This level of testing is inadequate for video game software, for insurance company website software, and for the software that runs your car radio. If a pull request with this lack of testing was submitted to the netCDF or HDF5 projects, it would be rejected out of hand as incomplete.

I understand that there has not been training here on unit testing, and we are arranging some. But unit testing has been common for more than 20 years, and is now ubiquitous. Every science software package we use has unit testing. The "lack of training" excuse is wearing pretty thin. 

Furthermore, proving the correctness of your work is something a scientist should already be familiar with. Yet this has not been done, or even attempted. Is this good science? It's definitely not good software engineering!

I do not mean to unfairly single out @mdtoy. These poor practices and careless habits are endemic to NOAA, and clearly point to a leadership problem. This is how we have tens of thousands of lines of untested UFS_UTILS code, to which we must now laboriously (and expensively) add the tests that the original programmers could not be bothered to write, and were not compelled to complete. This PR will add to that backlog of technical debt and unfinished programming work.

This PR has been open for months. I generally expect about the same amount of production code and test code. For 2500 lines of new code I expect about 2500 lines of test code. I would expect a decent programmer to write ~2500 lines of test code in a month or less. So in less time than it took to avoid, procrastinate, and evade this testing work, it could have been properly completed.

I do not believe that this code is generating correct science results. I would very much like to be proven wrong by a complete set of test codes that checked all functions and subroutines for correctness.

**Comment by GeorgeGayno-NOAA** [comment]
> I do not believe that this code is generating correct science results.
> 
> There are about 2500 lines of new Fortran code in this PR.
> 
> We (myself, @GeorgeGayno-NOAA , @mdtoy ) had a meeting to discuss the level of testing that this code has been subjected to:
> 1 - @mdtoy manually checked results at about eight different times during development.
> 2 - The regression test output was generated and scanned for infinities.
> 3 - The code was run within the model without causing any problems.
> 4 - The output was then accepted and now forms the basis of our regression tests.
> 
> We have no tests that check or demonstrate that the actual numbers generated are correct. We have no tests for boundary or error conditions.
> 
> This level of testing is inadequate for video game software, for insurance company website software, and for the software that runs your car radio. If a pull request with this lack of testing was submitted to the netCDF or HDF5 projects, it would be rejected out of hand as incomplete.
> 
> I understand that there has not been training here on unit testing, and we are arranging some. But unit testing has been common for more than 20 years, and is now ubiquitous. Every science software package we use has unit testing. The "lack of training" excuse is wearing pretty thin.
> 
> Furthermore, proving the correctness of your work is something a scientist should already be familiar with. Yet this has not been done, or even attempted. Is this good science? It's definitely not good software engineering!
> 
> I do not mean to unfairly single out @mdtoy. These poor practices and careless habits are endemic to NOAA, and clearly point to a leadership problem. This is how we have tens of thousands of lines of untested UFS_UTILS code, to which we must now laboriously (and expensively) add the tests that the original programmers could not be bothered to write, and were not compelled to complete. This PR will add to that backlog of technical debt and unfinished programming work.
> 
> This PR has been open for months. I generally expect about the same amount of production code and test code. For 2500 lines of new code I expect about 2500 lines of test code. I would expect a decent programmer to write ~2500 lines of test code in a month or less. So in less time than it took to avoid, procrastinate, and evade this testing work, it could have been properly completed.
> 
> I do not believe that this code is generating correct science results. I would very much like to be proven wrong by a complete set of test codes that checked all functions and subroutines for correctness.

@edwardhartnett I though we had agreed to merge the PR without tests for now.  Did you change your mind?

**Comment by edwardhartnett** [comment]
Merge if you feel that is best. I am raising concerns about the science results.

**Review by GeorgeGayno-NOAA — APPROVED** [review]
The branch at d94441f was successfully compiled on Orion, Jet, Hera, WCOSS-Cray and WCOSS-Dell.  The grid generation regression/consistency tests all passed.

**Review by JeffBeck-NOAA — APPROVED** [review]

**Comment by GeorgeGayno-NOAA** [comment]
> Merge if you feel that is best. I am raising concerns about the science results.

OK. But since you are a reviewer we can't merge until you 'approve'.

**Review by edwardhartnett — APPROVED** [review]

**Pull request merged**
